register-machines/src/main.rs

```diff
@@ -11,55 +11,203 @@ enum Instruction {
+            .fold(
+                (None as Option<u128>).encode(),
+                |x, y| Some((y, x)).encode()
+            )
+    }
+}
+
+trait Decodable {
+    fn decode(godel: Godel) -> Self;
+}
+
+impl Decodable for u128 {
+    fn decode(godel: Godel) -> Self {
+        godel
+    }
+}
+
+impl Decodable for usize {
+    fn decode(godel: Godel) -> Self {
+        godel as usize
+    }
 }
+
+impl<T: Decodable, V: Decodable> Decodable for (T, V) {
+    fn decode(godel: Godel) -> Self {
+        let x_godel = godel.trailing_ones() as Godel;
+        let y_godel = godel >> (x_godel + 1);
+
+        (decode(x_godel), decode(y_godel))
+    }
+}
+
+impl<T: Decodable> Decodable for Option<T> {
+    fn decode(godel: Godel) -> Self {
+        if godel == 0 {
+            None
+        } else {
+            Some(decode(godel - 1))
+        }
+    }
+}
+
+impl Decodable for Instruction {
+    fn decode(godel: Godel) -> Self {
+        match decode::<Option<(u128, Godel)>>(godel) {
+            None => Halt,
+            Some((f, arg)) => {
+                if f % 2 == 0 {
+                    // EVEN: add
+                    Add(f / 2, decode(arg))
+                } else {
+                    // ODD: sub
+                    let (j, k) = decode(arg);
+                    Sub((f - 1) / 2, j, k)
+                }
+            }
+        }
+    }
+}
+
+impl<T: Decodable> Decodable for Vec<T> {
+    fn decode(mut godel: Godel) -> Self {
+        let mut list = vec!();
+
+        while let Some((head, tail)) = decode(godel) {
+            list.push(head);
+
+            godel = tail;
+        }
+    
+        list
+    }
+}
+
+fn decode<T: Decodable>(godel: Godel) -> T {
+    T::decode(godel)
+}
+
+use Instruction::*;
+
+fn eval_program(program: &[Instruction], init: &State) -> State {
+    let (mut label, mut registers) = init.clone();
+
+    while label < program.len() {
+        /* Reveal effects of instruction to state. */
+        match program[label] {
+            Add(reg, l) => {
+                /* Increment target register. */
+                *registers.get_mut(&reg).unwrap() += 1;
+    
+                /* Jump to label. */
+                label = l;
+            },
+            Sub(reg, lsub, lnop) => {
+                let register = registers.get_mut(&reg).unwrap();
+    
+                if *register == 0 {
+                    /* Register can't be decremented. */
+                    label = lnop;
+                } else {
+                    /* Decrement register. */
+                    *register -= 1;
+                    label = lsub;
+                }
+            },
+            Halt => { break; },
+        };
+    }
+
+    return (label, registers);
+}
+
+fn encode_program_to_list(program: &[Instruction]) -> Vec<u128> {
+    program.iter().map(Encodable::encode).collect()
+}
+
 fn decode_list_to_program(program: &[u128]) -> Vec<Instruction> {
-    unimplemented!();
+    program.iter().map(|&g| decode(g)).collect()
```

Comment:
I like this, everything is inlined much like the other decode function.
Inlining is very fast, but may cause your binary to bloat if it happens too much. `cargo build --release` calls rustc with a compilation option `opt-level=3` (-O3 on C/C++ compilers). But there also exists `opt-level=s` and `opt-level=z` which optimise for binary size (z is more optimised for size than s). This is only useful info if you plan to work on microcontrollers, but I'd thought I'd let you know.